Switch to using distroless in Dockerfiles

Rithwik0480 opened the pull request:
switched to [Google's Distroless](https://github.com/GoogleContainerTools/distroless) for a resulting image in a multistage build. Distroless is a very minimal Docker image that tailored to only contain dependencies required for a particular programming language. For Rust it only contains glibc and its dependencies. See https://github.com/GoogleContainerTools/distroless/blob/main/cc/BUILD for how the base image is built. The resulting image does not contain a shell.

The resulting image is 40% smaller (128MB vs 76MB). 
fixes: #149

Rithwik0480 committed:
Update Dockerfile

Rithwik0480 committed:
Update Dockerfile

Rithwik0480 committed:
Update Dockerfile

Rithwik0480 committed:
Update Dockerfile

Rithwik0480 committed:
Update Dockerfile

smohan-dw commented on Dockerfile:
```diff
@@ -13,25 +13,17 @@ RUN cargo build --locked --profile ${PROFILE}
 
 # This is the 2nd stage: a very small image where we copy the Polkadot binary."
-FROM docker.io/library/ubuntu:20.04
+FROM gcr.io/distroless/cc-debian11@sha256:c2e1b5b0c64e3a44638e79246130d480ff09645d543d27e82ffd46a6e78a3ce3
```
https://console.cloud.google.com/gcr/images/distroless/global/cc-debian11@sha256:3516ad5504f54aeaea0444dfd5044cc5969653e49b594cecb47dc400ea6f6820/details - seems to be the latest

smohan-dw commented on Dockerfile:
```diff
@@ -13,25 +13,17 @@ RUN cargo build --locked --profile ${PROFILE}
 # RUN cargo test --release --all
 
 # This is the 2nd stage: a very small image where we copy the Polkadot binary."
```
CORD binary

Rithwik0480 committed:
Update Distroless image to latest

https://console.cloud.google.com/gcr/images/distroless/global/cc-debian11@sha256:3516ad5504f54aeaea0444dfd5044cc5969653e49b594cecb47dc400ea6f6820/details

Rithwik0480 committed:
Update Distroless image with :latest tag

Rithwik0480 reviewed (commented):
The distroless image used here is the image with `:latest` tag .

[https://console.cloud.google.com/gcr/images/distroless/global/cc-debian11@sha256:9b8e0854865dcaf49470b4ec305df45957020fbcf17b71eeb50ffd3bc5bf885d/details?tab=vulnz](distroless image)

smohan-dw commented:
CI is exiting with no space left in the device error.

2023-03-20-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-b78d27aa9e5e005b.rlib" "-Wl,-Bdynamic" "-lz" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/home/runner/.rustup/toolchains/nightly-2023-03-20-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/home/runner/work/cord/cord/target/debug/deps/inspect_works-ff0ee0df89a928aa" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs"
  = note: /usr/bin/ld: final link failed: No space left on device
          collect2: error: ld returned 1 exit status
